File: kline-backend/app/schemas/backtest.py
"""回测相关的 Pydantic 模型"""
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional


class BacktestRequest(BaseModel):
    """回测请求"""
    code: str = Field(..., description="股票代码")
    start_date: str = Field(..., description="回测开始日期 YYYY-MM-DD")
    end_date: str = Field(..., description="回测结束日期 YYYY-MM-DD")
    strategy_id: str = Field(..., description="策略ID")
    strategy_params: Dict[str, Any] = Field(default_factory=dict, description="策略参数")
    initial_capital: float = Field(100000.0, description="初始资金")
    position_ratio: float = Field(0.8, description="单次买入仓位比例（0-1）")


class TradeRecord(BaseModel):
    """交易记录"""
    date: str
    code: str
    name: str
    action: str
    price: float
    shares: int
    amount: float
    commission: float
    reason: str


class PositionInfo(BaseModel):
    """持仓信息"""
    code: str
    name: str
    shares: int
    avg_price: float
    current_price: float
    market_value: float
    cost: float
    profit: float
    profit_pct: float


class DailyPosition(BaseModel):
    """每日持仓"""
    date: str
    cash: float
    market_value: float
    total_value: float
    positions: List[PositionInfo]


class BacktestMetricsData(BaseModel):
    """回测绩效指标"""
    total_return: float = Field(..., description="总收益率(%)")
    annual_return: float = Field(..., description="年化收益率(%)")
    max_drawdown: float = Field(..., description="最大回撤(%)")
    sharpe_ratio: float = Field(..., description="夏普比率")
    win_rate: float = Field(..., description="胜率(%)")
    profit_loss_ratio: float = Field(..., description="盈亏比")
    total_trades: int = Field(..., description="总交易次数")
    win_trades: int = Field(..., description="盈利次数")
    loss_trades: int = Field(..., description="亏损次数")
    buy_hold_return: Optional[float] = Field(None, description="买入持有收益率(%)")
    excess_return: Optional[float] = Field(None, description="超额收益(%)")


class BacktestData(BaseModel):
    """回测结果数据"""
    stock_code: str
    stock_name: str
    start_date: str
    end_date: str
    strategy_name: str
    strategy_params: Dict[str, Any]
    initial_capital: float
    final_capital: float
    metrics: BacktestMetricsData
    daily_records: List[DailyPosition]
    trades: List[TradeRecord]
    equity_curve: List[Dict[str, Any]]  # 资金曲线 [{date, value}, ...]
    buy_hold_curve: Optional[List[Dict[str, Any]]] = None  # 买入持有基准曲线


class BacktestResponse(BaseModel):
    """回测响应"""
    code: int = 0
    message: str = "success"
    data: Optional[BacktestData] = None


class StrategyParam(BaseModel):
    """策略参数定义"""
    name: str
    label: str
    type: str
    default: Any
    min: Optional[float] = None
    max: Optional[float] = None
    step: Optional[float] = None


class StrategyDefinition(BaseModel):
    """策略定义"""
    id: str
    name: str
    description: str
    params: List[StrategyParam]


class StrategyListResponse(BaseModel):
    """策略列表响应"""
    code: int = 0
    message: str = "success"
    data: List[StrategyDefinition]

File: kline-frontend/src/types/backtest.ts
/**
 * 回测相关类型定义
 */

/**
 * 策略参数定义
 */
export interface StrategyParam {
  name: string;
  label: string;
  type: string;
  default: number;
  min?: number;
  max?: number;
  step?: number;
}

/**
 * 策略定义
 */
export interface StrategyDefinition {
  id: string;
  name: string;
  description: string;
  params: StrategyParam[];
}

/**
 * 回测请求
 */
export interface BacktestRequest {
  code: string;
  start_date: string;
  end_date: string;
  strategy_id: string;
  strategy_params: Record<string, number>;
  initial_capital: number;
  position_ratio: number;
}

/**
 * 交易记录
 */
export interface TradeRecord {
  date: string;
  code: string;
  name: string;
  action: 'buy' | 'sell';
  price: number;
  shares: number;
  amount: number;
  commission: number;
  reason: string;
}

/**
 * 持仓信息
 */
export interface PositionInfo {
  code: string;
  name: string;
  shares: number;
  avg_price: number;
  current_price: number;
  market_value: number;
  cost: number;
  profit: number;
  profit_pct: number;
}

/**
 * 每日持仓
 */
export interface DailyPosition {
  date: string;
  cash: number;
  market_value: number;
  total_value: number;
  positions: PositionInfo[];
}

/**
 * 回测绩效指标
 */
export interface BacktestMetrics {
  total_return: number;
  annual_return: number;
  max_drawdown: number;
  sharpe_ratio: number;
  win_rate: number;
  profit_loss_ratio: number;
  total_trades: number;
  win_trades: number;
  loss_trades: number;
  buy_hold_return?: number;
  excess_return?: number;
}

/**
 * 资金曲线数据点
 */
export interface EquityCurvePoint {
  date: string;
  value: number;
}

/**
 * 回测结果数据
 */
export interface BacktestData {
  stock_code: string;
  stock_name: string;
  start_date: string;
  end_date: string;
  strategy_name: string;
  strategy_params: Record<string, number>;
  initial_capital: number;
  final_capital: number;
  metrics: BacktestMetrics;
  daily_records: DailyPosition[];
  trades: TradeRecord[];
  equity_curve: EquityCurvePoint[];
  buy_hold_curve?: EquityCurvePoint[];
}

/**
 * 回测响应
 */
export interface BacktestResponse {
  code: number;
  message: string;
  data: BacktestData;
}

/**
 * 策略列表响应
 */
export interface StrategyListResponse {
  code: number;
  message: string;
  data: StrategyDefinition[];
}
